actionpack/lib/action_controller/metal/request_forgery_protection.rb

```diff
@@ -240,6 +288,56 @@ def handle_unverified_request
       end
     end
 
+    class SessionStore
+      def fetch(request)
+        request.session[:_csrf_token]
+      end
+
+      def store(request, csrf_token)
+        request.session[:_csrf_token] = csrf_token
+      end
+
+      def reset(request)
+        request.session.delete(:_csrf_token)
+      end
+    end
+
+    class CookieStore
+      def initialize(cookie = :csrf_token)
+        @cookie = cookie
+      end
+
+      def fetch(request)
+        contents = request.cookie_jar.encrypted[@cookie]
+        return nil if contents.nil?
+
+        value = JSON.parse(contents).symbolize_keys
+        return nil unless value&.fetch(:session_id) == request.session.id.to_s
+
+        value[:token]
+      end
+
+      def store(request, csrf_token)
+        request.cookie_jar.encrypted[@cookie] = {
+          value: {
+            token: csrf_token,
+            session_id: request.session.id.to_s,
+          }.to_json,
+          httponly: true,
+          same_site: :lax,
+          secure: true,
+        }
+      end
+
+      def reset(request)
+        request.cookie_jar.delete(@cookie)
+      end
+    end
+
+    def reset_csrf_token(request)
+      csrf_token_storage_strategy.reset(request)
+    end
```

Comment:
I feel weird about making this a public function, but I'm not sure what alternatives I have.